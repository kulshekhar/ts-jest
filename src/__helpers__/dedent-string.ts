export const omitLeadingWhitespace = (text: string): string => {
  return text.replace(/^\s+/gm, '')
}

export const dedent = (strings: TemplateStringsArray, ...values: unknown[]) => {
  let joinedString = ''
  for (let i = 0; i < values.length; i++) {
    joinedString += `${strings[i]}${values[i]}`
  }
  joinedString += strings[strings.length - 1]

  return omitLeadingWhitespace(joinedString)
}
