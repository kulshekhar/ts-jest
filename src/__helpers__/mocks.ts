import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
