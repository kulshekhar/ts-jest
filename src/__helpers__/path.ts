import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
