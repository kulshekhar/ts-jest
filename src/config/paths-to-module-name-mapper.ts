import { Config } from '@jest/types'
import { LogContexts } from 'bs-logger'
import { CompilerOptions } from 'typescript'

import { rootLogger } from '../util/logger'
import { Errors, interpolate } from '../util/messages'

type TsPathMapping = Exclude<CompilerOptions['paths'], undefined>
type JestPathMapping = Config.InitialOptions['moduleNameMapper']

// we don't need to escape all chars, so commented out is the real one
// const escapeRegex = (str: string) => str.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
const escapeRegex = (str: string) => str.replace(/[-\\^$*+?.()|[\]{}]/g, '\\$&')

const logger = rootLogger.child({ [LogContexts.namespace]: 'path-mapper' })

export const pathsToModuleNameMapper = (
  mapping: TsPathMapping,
  { prefix = '' }: { prefix?: string } = {},
): JestPathMapping => {
  const jestMap: JestPathMapping = {}
  for (const fromPath of Object.keys(mapping)) {
    let pattern: string
    const toPaths = mapping[fromPath]
    // check that we have only one target path
    if (toPaths.length === 0) {
      logger.warn(interpolate(Errors.NotMappingPathWithEmptyMap, { path: fromPath }))
      continue
    } else if (toPaths.length > 1) {
      logger.warn(
        interpolate(Errors.MappingOnlyFirstTargetOfPath, {
          path: fromPath,
          count: toPaths.length,
        }),
      )
    }
    const target = toPaths[0]

    // split with '*'
    const segments = fromPath.split(/\*/g)
    if (segments.length === 1) {
      pattern = `^${escapeRegex(fromPath)}$`
      jestMap[pattern] = `${prefix}${target}`
    } else if (segments.length === 2) {
      pattern = `^${escapeRegex(segments[0])}(.*)${escapeRegex(segments[1])}$`
      jestMap[pattern] = `${prefix}${target.replace(/\*/g, '$1')}`
    } else {
      logger.warn(interpolate(Errors.NotMappingMultiStarPath, { path: fromPath }))
      continue
    }
  }

  return jestMap
}
