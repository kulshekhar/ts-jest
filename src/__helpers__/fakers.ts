import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
