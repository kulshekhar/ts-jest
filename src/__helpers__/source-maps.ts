// WARNING: this file is shared between e2e and unit tests
import bufferFrom = require('buffer-from')
import stableStringify = require('fast-json-stable-stringify')
import { realpathSync } from 'fs'
import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

export function base64ToSourceMaps(base64: string): RawSourceMap {
  return JSON.parse(bufferFrom(base64, 'base64').toString('utf8'))
}

export function sourceMapsToBase64(sourceMaps: RawSourceMap): string {
  return bufferFrom(stableStringify(sourceMaps)).toString('base64')
}

export interface ParsedSourceWithMaps {
  sourceMaps?: RawSourceMap
  comment?: string
  source: string
}
export function parseSource(source: string): ParsedSourceWithMaps {
  const [comment, b64Maps]: [string, string | undefined] =
    // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
    (source.match(/[\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,(\S+)\s*$/) as any) ||
    []
  if (b64Maps) {
    const map = base64ToSourceMaps(b64Maps)
    return {
      source: source.substr(0, -comment.length),
      comment: comment.trim(),
      sourceMaps: map,
    }
  }
  return { source }
}

// source-map module doesn't provide a sync method to extract the source-maps
export function extractSourceMaps(source: string): RawSourceMap | undefined {
  return parseSource(source).sourceMaps
}

export function relativisePaths(map: RawSourceMap, fromPath: string, newPrefix = ''): RawSourceMap {
  const res = { ...map }
  const from = realpathSync(fromPath)
  const remap = (path: string): string =>
    (isAbsolute(path) ? `${newPrefix}${relative(from, realpathSync(path))}` : path).replace(/\\/g, '/')
  if (res.sourceRoot) res.sourceRoot = remap(res.sourceRoot)
  if (res.sources) res.sources = res.sources.map(remap)
  if (res.file) res.file = remap(res.file)
  return res
}

export function rewriteSourceMaps(source: string, sourceMapsTransformer: (maps: RawSourceMap) => RawSourceMap): string {
  return source.replace(
    /([\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,)(\S+)(\s*)$/,
    (_, before, base64, after) => {
      let map = base64ToSourceMaps(base64)
      map = sourceMapsTransformer(map)
      return `${before}${sourceMapsToBase64(map)}${after}`
    },
  )
}
