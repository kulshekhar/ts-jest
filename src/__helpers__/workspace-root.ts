import fs from 'node:fs'
import path from 'node:path'

function workspaceRootInner(dir: string, candidateRoot: string) {
  if (path.dirname(dir) === dir) return candidateRoot

  const matchablePaths = [path.join(dir, '.ts-jest-digest'), path.join(dir, '.github'), path.join(dir, 'renovate.json')]
  if (matchablePaths.some((x) => fs.existsSync(x))) {
    return dir
  }

  return workspaceRootInner(path.dirname(dir), candidateRoot)
}

export const workspaceRoot = workspaceRootInner(process.cwd(), process.cwd())
