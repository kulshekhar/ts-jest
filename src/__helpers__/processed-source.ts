import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

import { ROOT } from './path'
import { ParsedSourceWithMaps, parseSource, relativisePaths, rewriteSourceMaps } from './source-maps'

export default class ProcessedSource {
  readonly filename: string

  constructor(readonly output: string, filename: string, readonly cwd: string = ROOT) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)

    return undefined
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap => relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
