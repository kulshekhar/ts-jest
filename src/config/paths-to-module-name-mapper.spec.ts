import { logTargetMock } from '../__helpers__/mocks'

import { pathsToModuleNameMapper } from './paths-to-module-name-mapper'

const tsconfigMap = {
  log: ['src/utils/log'],
  server: ['src/server'],
  client: ['src/client', 'src/client/index'],
  'util/*': ['src/utils/*'],
  'api/*': ['src/api/*'],
  'test/*': ['test/*'],
  'mocks/*': ['test/mocks/*'],
  'test/*/mock': ['test/mocks/*', 'test/__mocks__/*'],
}

describe('pathsToModuleNameMapper', () => {
  it('should convert tsconfig mapping with no given prefix', () => {
    expect(pathsToModuleNameMapper(tsconfigMap)).toMatchInlineSnapshot(`
      Object {
        "^api/(.*)$": "src/api/$1",
        "^client$": Array [
          "src/client",
          "src/client/index",
        ],
        "^log$": "src/utils/log",
        "^mocks/(.*)$": "test/mocks/$1",
        "^server$": "src/server",
        "^test/(.*)$": "test/$1",
        "^test/(.*)/mock$": Array [
          "test/mocks/$1",
          "test/__mocks__/$1",
        ],
        "^util/(.*)$": "src/utils/$1",
      }
    `)
  })

  it.each(['<rootDir>/', 'foo'])('should convert tsconfig mapping with given prefix', (prefix) => {
    expect(pathsToModuleNameMapper(tsconfigMap, { prefix })).toMatchSnapshot(prefix)
  })

  it('should warn about mapping it cannot handle', () => {
    const log = logTargetMock()
    log.clear()
    expect(
      pathsToModuleNameMapper({
        kept: ['src/kept'],
        'no-target': [],
        'too/*/many/*/stars': ['to/*/many/*/stars'],
      }),
    ).toMatchInlineSnapshot(`
      Object {
        "^kept$": "src/kept",
      }
    `)
    expect(log.lines.warn).toMatchInlineSnapshot(`
      Array [
        "[level:40] Not mapping \\"no-target\\" because it has no target.
      ",
        "[level:40] Not mapping \\"too/*/many/*/stars\\" because it has more than one star (\`*\`).
      ",
      ]
    `)
  })
})
