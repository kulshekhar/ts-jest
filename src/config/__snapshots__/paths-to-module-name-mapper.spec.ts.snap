// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pathsToModuleNameMapper should convert tsconfig mapping with given prefix: <rootDir>/ 1`] = `
Object {
  "^@foo\\\\-bar/common$": "<rootDir>/../common/dist/library",
  "^api/(.*)$": "<rootDir>/src/api/$1",
  "^client$": Array [
    "<rootDir>/src/client",
    "<rootDir>/src/client/index",
  ],
  "^log$": "<rootDir>/src/utils/log",
  "^mocks/(.*)$": "<rootDir>/test/mocks/$1",
  "^server$": "<rootDir>/src/server",
  "^test/(.*)$": "<rootDir>/test/$1",
  "^test/(.*)/mock$": Array [
    "<rootDir>/test/mocks/$1",
    "<rootDir>/test/__mocks__/$1",
  ],
  "^util/(.*)$": "<rootDir>/src/utils/$1",
}
`;

exports[`pathsToModuleNameMapper should convert tsconfig mapping with given prefix: foo 1`] = `
Object {
  "^@foo\\\\-bar/common$": "foo/../common/dist/library",
  "^api/(.*)$": "foo/src/api/$1",
  "^client$": Array [
    "foo/src/client",
    "foo/src/client/index",
  ],
  "^log$": "foo/src/utils/log",
  "^mocks/(.*)$": "foo/test/mocks/$1",
  "^server$": "foo/src/server",
  "^test/(.*)$": "foo/test/$1",
  "^test/(.*)/mock$": Array [
    "foo/test/mocks/$1",
    "foo/test/__mocks__/$1",
  ],
  "^util/(.*)$": "foo/src/utils/$1",
}
`;
